File: install/src/main/java/org/jetbrains/bsp/bazel/install/EnvironmentCreator.java
package org.jetbrains.bsp.bazel.install;

import ch.epfl.scala.bsp4j.BspConnectionDetails;
import com.google.gson.GsonBuilder;
import io.vavr.control.Try;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import org.jetbrains.bsp.bazel.commons.Constants;

public class EnvironmentCreator {

  private final Path projectRootDir;
  private final BspConnectionDetails discoveryDetails;

  public EnvironmentCreator(Path projectRootDir, BspConnectionDetails discoveryDetails) {
    this.projectRootDir = projectRootDir;
    this.discoveryDetails = discoveryDetails;
  }

  public Try<Void> create() {
    return createDotBazelBsp().flatMap(__ -> createDotBsp());
  }

  private Try<Void> createDotBazelBsp() {
    return createDir(projectRootDir, Constants.DOT_BAZELBSP_DIR_NAME)
        .flatMap(this::createDotBazelBspFiles);
  }

  private Try<Void> createDotBazelBspFiles(Path dotBazelBspDir) {
    return copyAspects(dotBazelBspDir)
        .flatMap(__ -> createEmptyBuildFile(dotBazelBspDir))
        .flatMap(__ -> copyDefaultProjectViewFilePath(dotBazelBspDir));
  }

  private Try<Void> copyAspects(Path dotBazelBspDir) {
    var resourcesAspectsPath = "/" + Constants.ASPECTS_FILE_NAME;
    var destinationAspectsPath = dotBazelBspDir.resolve(Constants.ASPECTS_FILE_NAME);

    return copyFileFromResources(resourcesAspectsPath, destinationAspectsPath);
  }

  private Try<Void> createEmptyBuildFile(Path dotBazelBspDir) {
    var destinationBuildFilePath = dotBazelBspDir.resolve(Constants.BUILD_FILE_NAME);

    return Try.run(() -> destinationBuildFilePath.toFile().createNewFile());
  }

  private Try<Void> copyDefaultProjectViewFilePath(Path dotBazelBspDir) {
    var resourcesProjectViewFilePath = "/default-projectview.bazelproject";
    var destinationProjectViewFilePath = dotBazelBspDir.resolve("default-projectview.bazelproject");

    return copyFileFromResources(resourcesProjectViewFilePath, destinationProjectViewFilePath);
  }

  private Try<Void> copyFileFromResources(String resourcesPath, Path destinationPath) {
    return Try.withResources(() -> EnvironmentCreator.class.getResourceAsStream(resourcesPath))
        .of(inputStream -> copyFile(inputStream, destinationPath))
        .flatMap(x -> x);
  }

  private Try<Void> copyFile(InputStream inputStream, Path destinationPath) {
    return Try.run(
        () -> Files.copy(inputStream, destinationPath, StandardCopyOption.REPLACE_EXISTING));
  }

  private Try<Void> createDotBsp() {
    return createDir(projectRootDir, Constants.DOT_BSP_DIR_NAME)
        .flatMap(this::createBspDiscoveryDetailsFile);
  }

  private Try<Path> createDir(Path rootDir, String name) {
    var dir = rootDir.resolve(name);
    return Try.of(() -> Files.createDirectories(dir));
  }

  private Try<Void> createBspDiscoveryDetailsFile(Path dotBspDir) {
    var destinationBspDiscoveryFilePath = dotBspDir.resolve(Constants.BAZELBSP_JSON_FILE_NAME);
    var fileContent = new GsonBuilder().setPrettyPrinting().create().toJson(discoveryDetails);

    return writeStringToFile(destinationBspDiscoveryFilePath, fileContent);
  }

  private Try<Void> writeStringToFile(Path destinationPath, String string) {
    return Try.run(() -> Files.writeString(destinationPath, string));
  }
}

File: install/src/main/java/org/jetbrains/bsp/bazel/install/EnvironmentCreator.kt
package org.jetbrains.bsp.bazel.install

import ch.epfl.scala.bsp4j.BspConnectionDetails
import com.google.gson.GsonBuilder
import org.jetbrains.bsp.bazel.commons.Constants
import org.jetbrains.bsp.bazel.commons.flatMap
import org.jetbrains.bsp.bazel.commons.of
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.io.path.copyTo

abstract class EnvironmentCreator(private val projectRootDir: Path) {

    abstract fun create(): Result<Unit>

    protected fun createDotBazelBsp(): Result<Path> {
        removeOldDotBazelBspIfExists()

        val bazelBspDir = createDir(projectRootDir, Constants.DOT_BAZELBSP_DIR_NAME)
        return bazelBspDir.flatMap(::createDotBazelBspFiles).flatMap { bazelBspDir }
    }

    private fun removeOldDotBazelBspIfExists() {
        val oldBazelBspDir = projectRootDir.resolve(Constants.DOT_BAZELBSP_DIR_NAME)
        oldBazelBspDir.toFile().deleteRecursively()
    }

    private fun createDotBazelBspFiles(dotBazelBspDir: Path): Result<Unit> =
        Result.of { copyAspects(dotBazelBspDir) }.flatMap { createEmptyBuildFile(dotBazelBspDir) }

    private fun copyAspects(dotBazelBspDir: Path) {
        val destinationAspectsPath = dotBazelBspDir.resolve(Constants.ASPECTS_ROOT)
        copyAspectsFromResources("/" + Constants.ASPECTS_ROOT, destinationAspectsPath)
    }

    private fun createEmptyBuildFile(dotBazelBspDir: Path): Result<Unit> {
        val destinationBuildFilePath = dotBazelBspDir.resolve(Constants.BUILD_FILE_NAME)
        val destinationWorkspaceFilePath = dotBazelBspDir.resolve(Constants.WORKSPACE_FILE_NAME)
        return Result.of {
            destinationBuildFilePath.toFile().createNewFile()
            destinationWorkspaceFilePath.toFile().createNewFile()
        }
    }

    private fun copyAspectsFromResources(aspectsJarPath: String, destinationPath: Path) =
        javaClass.getResource(aspectsJarPath)?.let {
            val fileSystem = FileSystems.newFileSystem(it.toURI(), emptyMap<String, String>())
            copyFileTree(fileSystem.getPath(aspectsJarPath), destinationPath)
        } ?: error("Missing aspects resource")


    private fun copyFileTree(source: Path, destination: Path): Unit =
        Files.walk(source).forEach { copyUsingRelativePath(source, it, destination) }

    private fun copyUsingRelativePath(sourcePrefix: Path, source: Path, destination: Path) {
        val sourceRelativePath = sourcePrefix.relativize(source).toString()
        val destinationAbsolutePath = Paths.get(destination.toString(), sourceRelativePath)
        source.copyTo(destinationAbsolutePath, overwrite = true)
    }

    protected fun createDotBsp(discoveryDetails: BspConnectionDetails): Result<Unit> =
        createDir(projectRootDir, Constants.DOT_BSP_DIR_NAME)
            .flatMap { createBspDiscoveryDetailsFile(it, discoveryDetails) }

    protected fun createDir(rootDir: Path, name: String): Result<Path> {
        val dir = rootDir.resolve(name)

        return Result.of { Files.createDirectories(dir) }
    }

    private fun createBspDiscoveryDetailsFile(
        dotBspDir: Path,
        discoveryDetails: BspConnectionDetails
    ): Result<Unit> {
        val destinationBspDiscoveryFilePath = dotBspDir.resolve(Constants.BAZELBSP_JSON_FILE_NAME)
        return writeJsonToFile(destinationBspDiscoveryFilePath, discoveryDetails)
    }

    protected fun <T> writeJsonToFile(destinationPath: Path, data: T): Result<Unit> {
        val fileContent = GsonBuilder().setPrettyPrinting().create().toJson(data)
        return writeStringToFile(destinationPath, fileContent)
    }

    protected fun writeStringToFile(destinationPath: Path, string: String): Result<Unit> =
        Result.of { Files.writeString(destinationPath, string) }

}
